contributors/design-proposals/security-model.md

```diff
@@ -0,0 +1,190 @@
+# Requirements
+
+This document considers "customer facing" entities. For our purposes these are
+the Kubernetes API, and running workloads.
+
+1. Kubernetes needs to be safe in a multi-tenant environment
+2. Securing Kubernetes should be easy
+
+This document uses the "[CIA triad](https://en.wikipedia.org/wiki/Information_security#Key_concepts)" for analyzing these goals.
+
+# Kubernetes API access controls
+
+Kubernetes needs access controls capable of protecting the CIA of cluster
+resources in a multi-tenant environment.
+
+## Concerns with method based access controls
+
+Many Kubernetes deployments provide method based access control (e.g.,
+list pods). This is akin to securing \*nix with permissions on the "open"
+syscall. To allow more than privileged user, ACLs should be placed on nouns, not
+verbs.
+
+## Concerns with implicit relationships
+
+Kubernetes' implicit object relationships create subtle dangers.
+
+1. The "create pod" privilege allows a user to mount a secret, even if the user
+   cannot "get secrets". This is a failure of confidentiality.
+2. The "create pod" privilege can be used to place a pod behind a service, even
+   if the user cannot "update service". This is a failure of integrity.
+3. The "create pod" privilege can be used to place a pod behind a deployment,
+   which can be used to starve legitimate replicas. This is a failure of
+   integrity, and availability.
+
+This isn't a complete list of such dangers, but it's sufficient to show that
+access controls on an object are insufficient to guarantee CIA. Access controls
+are also needed on the relationships between objects.
+
+## Label ACLs (implicit ownership)
+
+Label ACLs are sometimes suggested as a method to provide access control on
+object-object relationships. Special permissions would be required to create or
+query an ACL'ed label. This approach is both complex and insufficient.
+
+1. The accidental omission of a label means that the system fails open.
+2. It becomes increasingly difficult to reason about permissions, especially if
+   set-based label selectors are used.
+3. Can ACLs be created on existing labels?
+   1. Yes: clients can be locked out of their workloads.
+   2. No: workloads must be restarted to change permissions.
+
+This approach also perturbs the existing behavior of label selectors. Does the
+"list pod" permission allow me to see pods with contentious ACLs? What about
+"list secrets"?
+
+Finally, this approach is insufficent, because not all relationships are defined
+by label selectors. Notably, secrets cannot be protected by label ACLs.
```

Comment:
The model being proposed is that access to secrets is gated by label.  Integrations do not have access to read all secrets, they only have access to read secrets by a particular label.  For instance, label `kubernetes.io/for-use-by-nodes=1`.  This would then not be special case - it would be normal case.  All "for use by" relationships can be modeled via ACL on labels for secrets.  

I do agree that targeted "-uses-" relationships are valuable, but only a small subset of those relationships matter, so I'm not sure that generic "uses" has significant value to users (I think it has to be proved that uses is a strictly better model than buckets + roles for the majority case, unless you're proposing to add "uses" as an additional dimension).  I also don't see that "-uses-" addresses confused deputy on controllers (unless I missed that in this doc), which is the more pressing concern.  And with controllers, if you grant a controller access to create anything that can use anything else, you also have to enable it to either transitively refer to all uses (thus requiring a fairly large set of permissions and reducing to the current model) or explicitly specify all uses (which for many users will result in them bulk selecting "all related resources" to grant access to).